mgr/manager.go

```diff
@@ -155,13 +158,112 @@ var validExtensions = map[string]bool{
 	".yml":  true,
 }
 
+// CheckDiskPKI checks the PKI information on disk against cert spec and alerts upon differences
+// Specifically, it checks that private key on disk matches spec algorithm & keysize,
+// and certificate  on disk matches CSR spec info
+func (m *Manager) CheckDiskPKI() error {
+
+	if m.Certs == nil || len(m.Certs) == 0 {
+		return nil
+	}
+
+	// Iterate through certificates. Compare what's on disk to what's structurally stored
+	for i := range m.Certs {
+		csm := m.Certs[i]
+		certPath := csm.Spec.Cert.Path
+		keyPath := csm.Spec.Key.Path
+		specPath := csm.Spec.Path
+		csrRequest := csm.Spec.Request
+
+		// Read private key algorithm and keysize from disk, determine if RSA or ECDSA
+		keyData, err := ioutil.ReadFile(keyPath)
+		if err != nil {
+			return err
+		}
+		pemKey, _ := pem.Decode(keyData)
+		if pemKey == nil {
+			return errors.New("Unable to pem decode private key on disk")
+		}
+
+		var algDisk string
+		var sizeDisk int
+		privKey, err := x509.ParsePKCS1PrivateKey(pemKey.Bytes)
+		if err != nil {
+			privKey, err := x509.ParseECPrivateKey(pemKey.Bytes)
+			if err != nil {
+				// If we get here, then invalid key type
+				return errors.New("manager: Unable to parse private key algorithm from disk")
+			}
+			// If we get here, then it's ECDSA
+			algDisk = "ecdsa"
+			sizeDisk = privKey.Curve.Params().BitSize
+		} else {
+			//If we get here, then it's RSA
+			algDisk = "rsa"
+			sizeDisk = privKey.N.BitLen()
+		}
+
+		// Check algorithm and keysize of private key on disk against what's defined in spec
+		algSpec := csrRequest.KeyRequest.Algo()
+		sizeSpec := csrRequest.KeyRequest.Size()
+
+		if algDisk != algSpec {
+			metrics.AlgorithmMismatchCount.WithLabelValues(specPath).Set(1)
+			fmt.Printf("ALERT: disk alg is %s but spec alg is %s\n", algDisk, algSpec)
+		} else {
+			metrics.AlgorithmMismatchCount.WithLabelValues(specPath).Set(0)
+		}
+
+		if sizeDisk != sizeSpec {
+			metrics.KeysizeMismatchCount.WithLabelValues(specPath).Set(1)
+			fmt.Printf("ALERT: disk key size is %d but spec key size is %d\n", sizeDisk, sizeSpec)
+		} else {
+			metrics.KeysizeMismatchCount.WithLabelValues(specPath).Set(0)
+		}
+
+		// Check that certificate hostnames match spec hostnames
+		certData, err := ioutil.ReadFile(certPath)
+		if err != nil {
+			return err
+		}
+		p, _ := pem.Decode(certData)
+		if p == nil {
+			return errors.New("Unable to pem decode certificate on disk")
+		}
+		cert, err := x509.ParseCertificate(p.Bytes)
+		if err != nil {
+			return err
+		}
+		if !reflect.DeepEqual(csrRequest.Hosts, cert.DNSNames) {
```

Comment:
It's generally recommended that unless you don't use `DeepEqual` outside of tests unless you have a good reason to, comparing your 2 `string[]`  yourself in a loop is much more performant - see https://stackoverflow.com/questions/15311969/checking-the-equality-of-two-slices#comment84006523_15312182

Reply:
Oh dang, I didn't realize that ordering still matters for DeepEquals. I don't really have a guarantee that the hosts and dnsnames will be in the same order -- so I'll write my own function sort then compare my string arrays. Thanks!